engines/experiment/plugin/rpc/factory.go

```diff
@@ -74,7 +82,44 @@ func (f *EngineFactory) GetExperimentRunner() (runner.ExperimentRunner, error) {
 	return f.runner, nil
 }
 
-func NewFactory(pluginBinary string, engineCfg json.RawMessage, logger *zap.SugaredLogger) (*EngineFactory, error) {
+func NewFactory(name string, cfg config.EngineConfig, logger *zap.SugaredLogger) (*EngineFactory, error) {
+	factoriesmu.Lock()
+	defer factoriesmu.Unlock()
+
+	// get a hash of the engine's configuration and use it as a configuration's fingerprint
+	cfgHash, err := hashstructure.Hash(cfg, hashstructure.FormatV2, nil)
```

Comment:
Is the name not enough to uniquely identify which experiment engine has been seen before? Why go through the trouble of hashing and comparing all properties?

Reply:
Actually, what's the intended behaviour if the `plugin_url` is the same across various routes, but if the other parameters in `strategy` are different?

Reply:
Hmm... I think from the config part, the `experiment_engine` is the engine name and `experiment_engine_properties` are the `cfg` (@romanwozniak CMIIW). And I don't think we will ever configure different `experiment_engine_properties` for the same `experiment_engine` which is why I wonder why the name alone cannot be used to determine whether an exp engine has been seen before (and therefore being initialised already).

```yaml
        experiment_engine: my-engine
        experiment_engine_properties:         
          plugin_url: http://plugins-server:8080/plugins/my-engine
          timeout: 1s
```

Reply:
> Actually, what's the intended behaviour if the plugin_url is the same across various routes, but if the other parameters in strategy are different?

Since `fiber.yml` is constructed by the Turing API, this scenario is not possible. But in case turing-router is deployed manually with a hand-crafted config file, then different instances of the `EngineFactory` should be created and that's the reason why we use a hash of the entire config struct.

Reply:
Hmm.. In that case, since the local download path is constructed like this: `filename := fmt.Sprintf("./%s", path.Base(downloadURL.Path))`, I'd think we should then limit to comparing this effective `filename` value (timeout shouldn't matter to re-trigger the download and not sure if the name does either). Anyway, not a big issue since routers created from our app wont encounter this.

Reply:
I simplified this example, in the real-world scenario you'll get more properties to compare:
```yaml
experiment_engine: my-engine
experiment_engine_properties:         
    plugin_url: http://plugins-server:8080/plugins/my-engine
    timeout: 1s
    key1: value1
    key2: value2
   ...
```

Since all engine configuration [is stored inside the factory object](https://github.com/romanwozniak/turing/blob/01bc1266067d1491578a8a2f6b2173cf0a4735e1/engines/experiment/plugin/rpc/factory.go#L32), we need to compare all the properties. Otherwise, in the situation when you have the same plugin, but configured with the different config values (only possible in the manual router deployment), then in both cases, the same factory object will be used, hence the experiment runner would not be properly configured.

But I think that in this scenario, it would be nice to download the plugin's binary under a different path to prevent conflicting downloads.